src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+                                               mir: &Mir<'tcx>,
+                                               source: MirSource) -> liveness::LocalSet {
+    use rustc_data_structures::indexed_set::IdxSetBuf;
+    let mut set = liveness::LocalSet::new_empty(mir.local_decls.len());
+    let result = liveness::liveness_of_locals(mir);
+    liveness::dump_mir(tcx, "generator_liveness", source, mir, &result);
+
+    for (block, data) in mir.basic_blocks().iter_enumerated() {
+        if let TerminatorKind::Yield { .. } = data.terminator().kind {
+            set.union(&result.outs[block]);
+        }
+    }
+
+    // The generator argument is ignored
+    set.remove(&Local::new(1));
+
+    set
+}
+
+fn compute_layout<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                            def_id: DefId,
+                            source: MirSource,
+                            interior: GeneratorInterior<'tcx>,
+                            mir: &mut Mir<'tcx>)
+    -> (HashMap<Local, (Ty<'tcx>, usize)>, GeneratorLayout<'tcx>)
+{
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let mut live_locals = locals_live_across_suspend_points(tcx, mir, source);
+
+    let allowed = tcx.erase_regions(&interior.as_slice());
+
+    for (local, decl) in mir.local_decls.iter_enumerated() {
+        if !live_locals.contains(&local) || decl.internal {
+            continue;
+        }
+        if !allowed.contains(&decl.ty) {
+            tcx.sess.span_warn(mir.span,
+                &format!("generator contains type {} in MIR, but typeck only knows about {}",
+                    decl.ty,
+                    interior));
+        }
+    }
+
+    let upvar_len = mir.upvar_decls.len();
+    let live_decls : Vec<_> = mir.local_decls
+        .iter_enumerated_mut()
+        .filter(|&(local, _)| live_locals.contains(&local))
+        .collect();
+
+    let mut remap = HashMap::new();
+    let unit = tcx.mk_nil();
+    let mut vars: Vec<_> = live_decls.into_iter().enumerate().map(|(idx, (local, decl))| {
+        let var = decl.clone();
+        *decl = LocalDecl {
+            mutability: Mutability::Mut,
+            ty: unit,
+            name: None,
+            source_info,
+            internal: false,
+            is_user_variable: false,
+        };
+        remap.insert(local, (var.ty, upvar_len + 1 + idx));
+        var
+    }).collect();
+
+    let layout = GeneratorLayout {
+        fields: vars
+    };
+
+    (remap, layout)
+}
+
+fn insert_entry_point<'tcx>(mir: &mut Mir<'tcx>,
+                            block: BasicBlockData<'tcx>) {
+    mir.basic_blocks_mut().raw.insert(0, block);
+
+    let blocks = mir.basic_blocks_mut().iter_mut();
+
+    for target in blocks.flat_map(|b| b.terminator_mut().successors_mut()) {
+        *target = BasicBlock::new(target.index() + 1);
+    }
+}
+
+fn elaborate_generator_drops<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
Yeah, drop elaboration will only drop upvars. This just replaces `drop(generator)` with the dropping of its upvars. Keeping `drop(generator)` around results in a recursive drop glue (since we're generating the drop glue here).

Do you have some alternative in mind?